src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    internal class RemoteBrowserFileStream : BrowserFileStream
+    {
+        private readonly IJSRuntime _jsRuntime;
+        private readonly ElementReference _inputFileElement;
+        private readonly int _maxSegmentSize;
+        private readonly PipeReader _pipeReader;
+        private readonly CancellationTokenSource _fillBufferCts;
+        private readonly TimeSpan _segmentFetchTimeout;
+
+        private bool _isReadingCompleted;
+        private bool _isDisposed;
+
+        public RemoteBrowserFileStream(
+            IJSRuntime jsRuntime,
+            ElementReference inputFileElement,
+            BrowserFile file,
+            RemoteBrowserFileStreamOptions options,
+            CancellationToken cancellationToken)
+            : base(file)
+        {
+            _jsRuntime = jsRuntime;
+            _inputFileElement = inputFileElement;
+            _maxSegmentSize = options.SegmentSize;
+            _segmentFetchTimeout = options.SegmentFetchTimeout;
+
+            var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: options.MaxBufferSize, resumeWriterThreshold: options.MaxBufferSize));
+            _pipeReader = pipe.Reader;
+            _fillBufferCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+            _ = FillBuffer(pipe.Writer, _fillBufferCts.Token);
+        }
+
+        private async Task FillBuffer(PipeWriter writer, CancellationToken cancellationToken)
+        {
+            long offset = 0;
+
+            while (offset < File.Size)
+            {
+                var pipeBuffer = writer.GetMemory(_maxSegmentSize);
+                var segmentSize = (int)Math.Min(_maxSegmentSize, File.Size - offset);
+
+                try
+                {
+                    using var readSegmentCts = CancellationTokenSource.CreateLinkedTokenSource(_fillBufferCts.Token);
```

Comment:
Nit: Either remove the `cancellationToken` parameter from `FillBuffer` or use it here:

```suggestion
                    using var readSegmentCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```